### PR TITLE
[Php84] Add rule for RoundingMode enum

### PR DESCRIPTION
**Pull request by jorgsowa**
Adds new rule for replacing mode constants in the function round() with RoundingMode cases in PHP 8.4.

The PR is not finished, but please let me know whether it makes sense to add such a rule to the PHP8.4 ruleset.

Referenced RFC: https://wiki.php.net/rfc/correctly_name_the_rounding_mode_and_make_it_an_enum

### CODE REVIEW

File: rules/Php84/Rector/FuncCall/RoundingModeEnumRector.php [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                    'PHP_ROUND_HALF_EVEN' => 'HalfEven',
+                    'PHP_ROUND_HALF_ODD' => 'HalfOdd',
+                    default => null,
+                };
+
+                if ($enumCase === null) {
+                    return null;
+                }
+
+                $args[2]->value = new Node\Expr\ClassConstFetch(new Name('RoundingMode'), $enumCase);
```
Comment (samsonasik) [inline rules/Php84/Rector/FuncCall/RoundingModeEnumRector.php]: use `FullyQualified` node to ensure class reference is valid, make shortname is auto import task

File: rules-tests/Php84/Rector/FuncCall/RoundingModeEnumRector/Fixture/do_not_reprint_node_type.php.inc [hunk 2]
```diff
@@ -0,0 +1,25 @@
+<?php
+
```
Comment (samsonasik) [inline rules-tests/Php84/Rector/FuncCall/RoundingModeEnumRector/Fixture/do_not_reprint_node_type.php.inc]: add Fixture namespace when possible

File: rules/Php84/Rector/FuncCall/RoundingModeEnumRector.php [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    /**
+     * @param FuncCall $node
+     */
+    public function refactor(Node $node): ?Node\Expr\FuncCall
+    {
+
+        if (!$this->isName($node, 'round')) {
+            return null;
+        }
+
```
Comment (samsonasik) [inline rules/Php84/Rector/FuncCall/RoundingModeEnumRector.php:57]: skip first class callable:

```php
if ($node->isFirstClassCallable()) {
    return null;
}

Comment (jorgsowa) [reply rules/Php84/Rector/FuncCall/RoundingModeEnumRector.php:57]: Done

File: rules-tests/Php84/Rector/FuncCall/RoundingModeEnumRector/Fixture/do_not_reprint_node_type.php.inc (hunk 2 above)
Comment (jorgsowa) [reply rules-tests/Php84/Rector/FuncCall/RoundingModeEnumRector/Fixture/do_not_reprint_node_type.php.inc]: Done

File: rules/Php84/Rector/FuncCall/RoundingModeEnumRector.php (hunk 1 above)
Comment (jorgsowa) [reply rules/Php84/Rector/FuncCall/RoundingModeEnumRector.php]: Done

**Comment by samsonasik** [comment]
Just a quick read, if I understand correctly, the constant is not deprecated (yet) in php 8.4, it just replaced in core code, so I think it is not really needed for now.

https://3v4l.org/Z40Kp/rfc#vgit.master

when on future, the constant will be deprecated, eg: on php 8.5, we can create the rule to migrate it, if this is accepted, we need to prepare immediatelly the Downgrade `RoundingMode` rules.

**Comment by jorgsowa** [comment]
It's unlikely that constants will be deprecated soon as it was the result of the RFC discussion. That's why I thought such a rule would help to speed up this process.

**Comment by samsonasik** [comment]
I prefer to have downgrade rule instead first then, as replace constant is not required, but once it upgraded to enum, it needs to have downgrade rule to back to constant

https://github.com/rectorphp/rector-downgrade-php

**Comment by jorgsowa** [comment]
Makes sense. I will create a downgrade rule then.

File: rules/Php84/Rector/FuncCall/RoundingModeEnumRector.php [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                    'PHP_ROUND_HALF_EVEN' => 'HalfEven',
+                    'PHP_ROUND_HALF_ODD' => 'HalfOdd',
+                    default => null,
+                };
+
+                if ($enumCase === null) {
+                    return null;
+                }
+
+                $args[2]->value = new Node\Expr\ClassConstFetch(new FullyQualified('RoundingMode'), $enumCase);
```
Comment (samsonasik) [inline rules/Php84/Rector/FuncCall/RoundingModeEnumRector.php]: add `$hasChanged` flag here, so later can return `null` when no constant passed

```suggestion
                $args[2]->value = new Node\Expr\ClassConstFetch(new FullyQualified('RoundingMode'), $enumCase);
                $hasChanged = true;
```

with init first before if above:

```php
        $hasChanged = false;
        if ($modeArg instanceof ConstFetch) {
```

then later, before return node, it check `$hasChanged` flag:

```php
if ($hasChanged) {
    return $node;
}

return null;
```

**Comment by samsonasik** [comment]
Thank you @jorgsowa

**Review by samsonasik — APPROVED** [review]

**Pull request merged**